src/coreclr/System.Private.CoreLib/src/System/GC.CoreCLR.cs

```diff
@@ -328,14 +328,20 @@ public static void SuppressFinalize(object obj)
         // for which SuppressFinalize has already been called. The other situation
         // where calling ReRegisterForFinalize is useful is inside a finalizer that
         // needs to resurrect itself or an object that it references.
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        private static extern void _ReRegisterForFinalize(object o);
+        [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "GCInterface_ReRegisterForFinalize")]
+        private static partial void ReRegisterForFinalize(ObjectHandleOnStack o);
 
-        public static void ReRegisterForFinalize(object obj)
+        public static unsafe void ReRegisterForFinalize(object obj)
         {
             ArgumentNullException.ThrowIfNull(obj);
 
-            _ReRegisterForFinalize(obj);
+            MethodTable* pMT = RuntimeHelpers.GetMethodTable(obj);
+            if (pMT->HasFinalizer)
+            {
+                ReRegisterForFinalize(ObjectHandleOnStack.Create(ref obj));
+            }
+
+            KeepAlive(obj);
```

Comment:
Is this required? The object should be kept alive by `ObjectHandleOnStack`.

Reply:
Other similar places have a short comment that explains why it is omitted. I think it would be more appropriate in this case.

Reply:
> For what it's worth, taking it out did not change the generated X64 assembly on Linux.

`KeepAlive` may also affect GC info reported by JIT. It's better to view dasm with GC info in this case (I forgot the environment variable).